src/provider/mlkzg.rs

```diff
@@ -0,0 +1,621 @@
+    ck.trim(ck.length() - 1)
+  }
+
+  fn prove(
+    ck: &UniversalKZGParam<E>,
+    _pk: &Self::ProverKey,
+    transcript: &mut <NE as NovaEngine>::TE,
+    C: &Commitment<NE>,
+    hat_P: &[E::Fr],
+    point: &[E::Fr],
+    eval: &E::Fr,
+  ) -> Result<Self::EvaluationArgument, NovaError> {
+    let x: Vec<E::Fr> = point.to_vec();
+
+    //////////////// begin helper closures //////////
+    let kzg_open = |f: &[E::Fr], u: E::Fr| -> E::G1Affine {
+      // On input f(x) and u compute the witness polynomial used to prove
+      // that f(u) = v. The main part of this is to compute the
+      // division (f(x) - f(u)) / (x - u), but we don't use a general
+      // division algorithm, we make use of the fact that the division
+      // never has a remainder, and that the denominator is always a linear
+      // polynomial. The cost is (d-1) mults + (d-1) adds in E::Scalar, where
+      // d is the degree of f.
+      //
+      // We use the fact that if we compute the quotient of f(x)/(x-u),
+      // there will be a remainder, but it'll be v = f(u).  Put another way
+      // the quotient of f(x)/(x-u) and (f(x) - f(v))/(x-u) is the
+      // same.  One advantage is that computing f(u) could be decoupled
+      // from kzg_open, it could be done later or separate from computing W.
+
+      let compute_witness_polynomial = |f: &[E::Fr], u: E::Fr| -> Vec<E::Fr> {
+        let d = f.len();
+
+        // Compute h(x) = f(x)/(x - u)
+        let mut h = vec![E::Fr::ZERO; d];
+        for i in (1..d).rev() {
+          h[i - 1] = f[i] + h[i] * u;
+        }
+
+        h
+      };
+
+      let h = compute_witness_polynomial(f, u);
+
+      <NE::CE as CommitmentEngineTrait<NE>>::commit(ck, &h)
+        .comm
+        .preprocessed()
+    };
+
+    let kzg_open_batch = |C: &[E::G1Affine],
+                          f: &[Vec<E::Fr>],
+                          u: &[E::Fr],
+                          transcript: &mut <NE as NovaEngine>::TE|
+     -> (Vec<E::G1Affine>, Vec<Vec<E::Fr>>) {
+      let poly_eval = |f: &[E::Fr], u: E::Fr| -> E::Fr {
+        // Horner's scheme
+        let mut v = f[f.len() - 1];
+
+        for fi in f.iter().rev().skip(1) {
+          v *= u;
+          v += fi;
+        }
+
+        v
+      };
+
+      let scalar_vector_muladd = |a: &mut Vec<E::Fr>, v: &Vec<E::Fr>, s: E::Fr| {
+        assert!(a.len() >= v.len());
+        for i in 0..v.len() {
+          a[i] += s * v[i];
+        }
+      };
+
+      let kzg_compute_batch_polynomial = |f: &[Vec<E::Fr>], q: E::Fr| -> Vec<E::Fr> {
+        let k = f.len(); // Number of polynomials we're batching
+
+        let q_powers = Self::batch_challenge_powers(q, k);
+
+        // Compute B(x) = f[0] + q*f[1] + q^2 * f[2] + ... q^(k-1) * f[k-1]
+        let mut B = f[0].clone();
+        for i in 1..k {
+          scalar_vector_muladd(&mut B, &f[i], q_powers[i]); // B += q_powers[i] * f[i]
+        }
+
+        B
+      };
```

Comment:
We should have a higher level functionality for computing the random-linear combination of vectors.

Reply:
see #223